### PR TITLE
621 confirming emails when creating accounts

### LINKED ISSUE #621: Confirming Emails When Creating Accounts
## **Expected Behavior:**

When a user creates an account have an email sent to them with a confirmation link that will flag their account as `confirmed`.

### Frontend:
- [x] Change the account creation page to redirect users to their inbox to confirm their emails.
- [x] Create an "Email Confirmed" page that will take the token from the URL and flag their account as confirmed and then take them to the sign in page.

### Backend:
- [x] After sign up, create a temporary token and email it to the user in the form of a URL. You can use `JWT` or `JSON Web Tokens` and the `EmailServices.js` to send emails. 
- [x] Change the sign in routes to stop users without confirmed emails to log in. You will need to add a `confirmUser()` function to `UserServices.js` and edit the `UserModel.js` to keep a confirmed flag

## **External Factors:**

Describe any external requirements that might effect the issue (e.g need to await information from other party).

## **Related Issues/PRs:**
- [Look at the password reset procedure](https://github.com/UofT-Frosh-Orientation/orientation-website/issues/213) [Pull Request](https://github.com/UofT-Frosh-Orientation/orientation-website/pull/283/files)
## **Resources:**
Contact @Freeassassin if you need help with working with the backend.
### **Tutorials:**

- [JWT Tutorial](https://youtu.be/4LLGarqzsxQ)
- [35 Minute Express.js Tutorial](https://www.youtube.com/watch?v=SccSCuHhOw0)
- [REST API Explainer Video through Express.js](https://www.youtube.com/watch?v=-MTSQjw5DrM)
- [Middleware Explainer Video](https://www.youtube.com/watch?v=lY6icfhap2o)
### **Documentation:**

- [JSON Web Token Documentation](https://www.npmjs.com/package/jsonwebtoken)
- [Internal Backend Documentation](https://docs.orientation.skule.ca/pages/Onboarding/Backend.html)

### PR DESCRIPTION
**Pull request by uzFer**
### **Issue:**
Closes #621 

### **Accomplishments:**
- Edited Sign Up page to tell users to check inbox for a verification link.
- Added 'confirmed' flag in `UserModel.js` which checks if they verified their account through email token or not
- Sends email token upon signing up using EmailServices.js. Instead of making a new function in `UserServices.js,` I tried to  update createUser so that an email is sent after creating an account. 
- Updated `userRoutes.js` with a route to confirmUser in `UserController.js`. This employs the validateEmailConfirmationToken function in `UserServices.js` to check the token and change the flag accordingly.
- Included another condition to the login function in `UserController.js` so that unconfirmed users should not be able to log in
- Created new page called `EmailConfirmed.jsx` to display status after the email token is sent. This should show a 'Perfect' or 'Oh no' message and have a button to let users log in again.

### **Visuals:** 
(I'll update once Docker works again)
**Updated Sign up Page:**
image
**Verification Complete:**
image
**Error message for incorrect token:** 
image
**Simple Email with URL**
image

### **Expected Behavior:**
- You should be able to sign up normally. The message after submitting the sign up form should tell you to check your inbox.
- An email with a URL should be sent that you can click to verify your account.
- Your account should then be confirmed if done correctly and you will be prompted to log in again. Otherwise, you will get an 'Oh no' message and have a button to try signing up again.

### **Testing Steps:**
1.  Set `readyForFrosh = true`
2. Run the backend using `docker-compose --up build`
3. Head to the login page --> Create account
4. Put in valid email address and info, check your inbox, click the link, and see if your account was confirmed correctly. You can also put invalid ones and check if you are able to login, too.

### **Complications:**
- Had some problems with committing, I need to check if it added the right files. Also, I did use `git config --unset core.hooksPath` to 'fix' the issue, not sure if this interferes with anything in dev.
- Right after signup, the user is automatically logged in. If they go through with verification they have to log in again, but not sure how to stop the initial one
- Wasn't able to use a separete .env token, so I just used the reset one to test. Let me know if there's a workaround.
- I have no clue how to use the template email so I just used a simple one.

### **Resources:**
- For my committing issues: [(https://github.com/typicode/husky/issues/1038)]

### CODE REVIEW

**Comment by Tanushanth** [comment]
Works perfectly, I tested with the frosh account and Leedur accounts, and the email comes through correctly. 
Just one small thing which I'm not sure how it would be implemented, but the first time when you sign up and it tells you to check your email, you are already logged in. Afterward, it won't let you in without verifying so thats fine.
Other than that it's all good!

**Review by Tanushanth — APPROVED** [review]
Works perfectly, I tested with the frosh account and Leedur accounts, and the email comes through correctly.
Just one small thing which I'm not sure how it would be implemented, but the first time when you sign up and it tells you to check your email, you are already logged in. Afterward, it won't let you in without verifying so thats fine.
Other than that it's all good!

File: client/src/pages/EmailConfirmed/EmailConfirmed.jsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    try {
+      const emailAndEmailToken = {
+        email: email,
+        emailToken: emailToken,
+      };
+      axios.post('/user/verify-user-email', emailAndEmailToken);
+      setvalidEmailToken(true);
+    }
+    catch(e) {
+      console.log(e);
```
Comment (Freeassassin) [inline client/src/pages/EmailConfirmed/EmailConfirmed.jsx]: Dont use console.log() here, instead if it is a user facing error use the `setSnackbar` component.

File: client/src/pages/SignUp/SignUp.jsx [hunk 2]
```diff
@@ -252,19 +252,25 @@ const PageSignUp = () => {
...
+                </h3>
+                {/* <Link to="/registration" className="no-link-style">
                   <div>
                     <Button
                       label="Register"
                       style={{ padding: '25px 60px', fontSize: '20px', borderRadius: '20px' }}
                     />
                   </div>
-                </Link>
+                </Link> */}
```
Comment (Freeassassin) [inline client/src/pages/SignUp/SignUp.jsx]: Remove extra comments.

File: server/src/controllers/UserController.js [hunk 3]
```diff
@@ -39,17 +39,18 @@ const UserController = {
...
-        if (err) {
-          return next(err);
-        }
+      //req.logIn(user, (err) => {
+        //if (err) {
+        //  return next(err);
+        //}
         return res.status(200).send({ message: 'Success!', user: user.getResponseObject() });
-      });
+      //});
```
Comment (Freeassassin) [inline server/src/controllers/UserController.js]: Remove extra comments.

**Review by Freeassassin — CHANGES_REQUESTED** [review]
Hey Uzma, overall great work! The functionality is all there and everything works properly. However, there are a couple of things I need you to take care of:

- [x] Currently after you sign up you still get logged in automatically. This is because of how sign-up works in the `/state/user/saga.jsx` file. Edit the files in `state/user/` to separate signup from sign in by adding to the `useSlice.jsx` the following `singupStart`, `singupSuccess`, `singupFail`. 
- [x] In the backend instead of sending the email in services files do it through the `newUserSubcription.js`. This will make sure that the backend will be able to send a response and take care of the email later. For reference look at how it is done in `requestPasswordReset()`. Additionally, make sure that the sign up email is sent after the email confirmation has been completed.  


Feel free to reach for more clarification.

**Comment by Tanushanth** [comment]
Tested again, it works one after the other now.

File: server/src/services/LeadurServices.js [hunk 4]
```diff
@@ -1,4 +1,6 @@
 const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const EmailServices = require('./EmailServices');
 
 const LeadurModel = require('../models/LeadurModel');
 const newUserSubscription = require('../subscribers/newUserSubscription');
```
Comment (Freeassassin) [inline server/src/services/LeadurServices.js]: Get rid of the unused imports.
```suggestion
const LeadurModel = require('../models/LeadurModel');
```

Remove unused imports.

**Review by Freeassassin — APPROVED** [review]
Hey Uzma, everything is great! Thank you for your time and effort on this issue! Just make sure to take care of your merge conflicts 👍

**Pull request merged**